transports/webrtc/src/udp_mux.rs

```diff
@@ -0,0 +1,382 @@
+// MIT License
+//
+// Copyright (c) 2021 WebRTC.rs
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+use async_trait::async_trait;
+use stun::{
+    attributes::ATTR_USERNAME,
+    message::{is_message as is_stun_message, Message as STUNMessage},
+};
+use tokio_crate as tokio;
+use tokio_crate::sync::Mutex;
+use webrtc_ice::udp_mux::{UDPMux, UDPMuxConn, UDPMuxConnParams, UDPMuxWriter};
+use webrtc_util::{sync::RwLock, Conn, Error};
+
+use std::{
+    collections::{HashMap, HashSet},
+    io::ErrorKind,
+    net::SocketAddr,
+    sync::{
+        atomic::{AtomicBool, Ordering},
+        Arc, Weak,
+    },
+};
+
+const RECEIVE_MTU: usize = 8192;
+
+/// A previously unseen address of a remote who've sent us an ICE binding request.
+#[derive(Debug)]
+pub struct NewAddr {
+    pub addr: SocketAddr,
+    pub ufrag: String,
+}
+
+/// Parameters for [`UDPMuxNewAddr`].
+pub struct UDPMuxParams {
+    conn: Box<dyn Conn + Send + Sync>,
+}
+
+impl UDPMuxParams {
+    /// Creates new params.
+    pub fn new<C>(conn: C) -> Self
+    where
+        C: Conn + Send + Sync + 'static,
+    {
+        Self {
+            conn: Box::new(conn),
+        }
+    }
+}
```

Comment:
We are constructing this from a regular UdpSocket so we can also ditch the `Conn` abstraction here which should give us access to `poll`-based APIs provided by tokio's UDP socket.

Reply:
I'm going to give it a try.